### DISCUSSION TITLE
Week 6 - Our current architecture

### DISCUSSION THREAD

**Discussion by WeberWr** [The-Escape-Box/TheEscapeBox#32] [General]
Welcome back to our weekly blog!

In this blog we want to introduce you a bit in our current game architecture. To show you that, we decided to focus on the first map scene. The architecture for the other scenes are not really complex and its not needed to give you an overview about what we have.

## Our structure
![Architektur-TheEscpaeBox](https://github.com/The-Escape-Box/TheEscapeBox/assets/97832247/51758221-d686-48ba-b25a-af344f48ab13)

## The handlers
In the center if our architecture are our three singleton handlers. We need them to handle all the stuff we need to access everywhere and all the stuff we need exactly once in our game. We know, that a single game handler would be enough, but we like the idea of single responsibility and so we decided to split it up to: 

- the PlayerHealthHandler that deals handles the health of the player, has control over the health bar and also gets imported in the upgrade service to deal with the heal. We wanted a singleton as handler here, because we don't want to import the player-gameobject in the other components and the player should not be responsible for the UI.
- the AmmunitionHandler that controls the ammunition counter in the UI and is imported in the gun and the ammunition-restock upgrade. 
- and the BloodBankHandler that is basically our money system. It holds the current blood-value of the player.

## Single-responsibility
Aside of the handlers we have five bigger parts in our architecture. In all of this we tried to follow the Single-Responsibility-Principle as good as we can. Out of this follows, that we have one script for every part of logic in the game and every script only deals with a small part of logic. As example you can look at the Enemy.  

<img width="242" alt="Enemy" src="https://github.com/The-Escape-Box/TheEscapeBox/assets/97832247/2e1fa756-8d52-4ce0-bc12-c957ac61e35b">

The enemy basically has four scripts. One for the sound, one for the attack, one for the movement and one for the health. The animations has no own script, because its only like two lines of code in these scripts, but aside from the animations every script only deals with one "feature" of the enemy.

## The puzzle
<img width="182" alt="Puzzle" src="https://github.com/The-Escape-Box/TheEscapeBox/assets/97832247/a86fc279-d954-4453-b777-c89efd7030b3">

Without any real dependencies there is the puzzle handling. Its a part for itself and stand fully alone. The special thing here is, that we do not have a specific handler for the puzzle. We decided to bind the puzzle to the gameobject of the transformer-box we introduced two weeks ago. We decided to do that, because the puzzle is hard connected to the transformer-box and if we decide to do a new puzzle we also need a new puzzle script to handle the different steps of the puzzle. The hard connection of the puzzle to the gameobject is necessary, while the player can only have a loose connection to its health, because the player itself don't have to worry about it health.

## What else is new in the game?
Of course the structure is not the only thing we did last week. We worked on a lots of small stuff and with focus in optimizing the game experience in the second map we introduced you last week. Be aware to check out the blog next week (maybe there comes a new playable version for you). 

## Time spend
Time we spent this week:
Mohamad: 7h
Lukas: 25h

The last thing to say is keep on giving us your feedback. We always like to see your ideas and suggestions for improvements. 

So game on! Farewell and see you at the next update! Remember, in the world of game development, bugs are just unexpected features waiting to be discovered... or squashed!

**Comment by tonoizer** [comment]
Great insights into your game architecture! I'm impressed by how you've structured the handlers and components with a focus on single responsibility. Looking forward to seeing more updates on your progress. Keep up the good work!
